backend/analyze_tailcall_graph.ml

```diff
@@ -492,12 +508,202 @@ end = struct
+            { (* Invariant: `t.phase = Uses_unknown_edge` iff for some `e` in
+                 `edges` this priority is associated with, `e.from =
+                 Unknown_fn`. *)
+              phase : phase;
+              (* Invariant: `t.length = List.length edges` *)
+              length : int
+            }
+
+          let phase t = t.phase
+
+          (* Compare candidate paths lexiographically by (phase, length). Order
+             paths that don't use an unknown edge earlier. *)
+          let compare fst snd =
+            let phase_to_ord = function
+              | Doesn't_use_unknown_edge -> 0
+              | Uses_unknown_edge -> 1
+            in
+            let c =
+              Int.compare (phase_to_ord fst.phase) (phase_to_ord snd.phase)
+            in
+            if c <> 0 then c else Int.compare fst.length snd.length
+        end
+
+        include T
+
+        module Pqueue = Regalloc_gi_utils.Make_min_priority_queue (struct
+          include T
+
+          let to_string _ = "dummy unused"
+        end)
+      end
+
+      type t =
+        { priority : Priority.t;
+          edges : Edge.With_loc.t list (* Nonempty list *)
+        }
+
+      let priority (t : t) = t.priority
+
+      let edges (t : t) = t.edges
+
+      let tip (t : t) = (List.hd t.edges).edge.to_
+
+      let singleton (with_loc : Edge.With_loc.t) : t =
+        let phase : Priority.phase =
+          match with_loc.edge.from with
+          | Unknown_fn -> Uses_unknown_edge
+          | Known_fn _ -> Doesn't_use_unknown_edge
+        in
+        { priority = { phase; length = 1 }; edges = [with_loc] }
+
+      let cons (with_loc : Edge.With_loc.t) ({ priority; edges } : t) : t =
+        let phase : Priority.phase =
+          match with_loc.edge.from with
+          | Unknown_fn -> Uses_unknown_edge
+          | Known_fn _ -> priority.phase
+        in
+        { priority = { phase; length = priority.length + 1 };
+          edges = with_loc :: edges
+        }
+    end
+
+    (* Gas is decremented once per edge examined. *)
+    let find_cycle_with_edge graph ~(with_loc : Edge.With_loc.t) ~gas :
+        (Edge.With_loc.t list * bool) option * int =
+      let module Priority = Tendril.Priority in
+      let visited = Vertex.Hashset.create 100 in
+      let enqueue, dequeue =
+        let q : Tendril.t Priority.Pqueue.t =
+          Priority.Pqueue.make ~initial_capacity:100
+        in
+        let enqueue tendril =
+          Priority.Pqueue.add q ~priority:(Tendril.priority tendril)
+            ~data:tendril
+        in
+        let dequeue () =
+          if Priority.Pqueue.is_empty q
+          then None
+          else
+            let elt = Priority.Pqueue.get_and_remove q in
+            Some (elt.priority, elt.data)
+        in
+        enqueue, dequeue
+      in
+      let source = with_loc.edge.from in
+      Vertex.Hashset.add visited source;
+      enqueue (Tendril.singleton with_loc);
+      let rec loop gas =
```

Comment:
I think this is a case where I would have preferred an appropriately-scoped ref. It's true we often look for ways to avoid mutable state, but threading through `gas` here I think just makes the code more complicated without increasing my confidence in its correctness (for example, making sure that we didn't forget to thread the updated `gas` through somewhere requires some mental overhead).

Reply:
Addressed.